libnestutil/block_vector.h

```diff
@@ -0,0 +1,642 @@
+{
+}
+
+template < typename value_type_ >
+inline BlockVector< value_type_ >::~BlockVector() = default;
+
+template < typename value_type_ >
+inline value_type_& BlockVector< value_type_ >::operator[]( const size_t pos )
+{
+  // Using bitwise operations to efficiently map the index to the
+  // right block and element.
+  const auto block_index = pos >> block_size_shift;
+  const auto element_index = pos & max_block_size_sub_1;
+  return blockmap_[ block_index ][ element_index ];
+}
+
+template < typename value_type_ >
+inline const value_type_& BlockVector< value_type_ >::operator[](
+  const size_t pos ) const
+{
+  // Using bitwise operations to efficiently map the index to the
+  // right block and element.
+  const auto block_index = pos >> block_size_shift;
+  const auto element_index = pos & max_block_size_sub_1;
+  return blockmap_[ block_index ][ element_index ];
+}
+
+template < typename value_type_ >
+inline typename BlockVector< value_type_ >::iterator
+BlockVector< value_type_ >::begin()
+{
+  return iterator( *this );
+}
+
+template < typename value_type_ >
+inline typename BlockVector< value_type_ >::const_iterator
+BlockVector< value_type_ >::begin() const
+{
+  return const_iterator( *this );
+}
+
+template < typename value_type_ >
+inline typename BlockVector< value_type_ >::iterator
+BlockVector< value_type_ >::end()
+{
+  return finish_;
+}
+
+template < typename value_type_ >
+inline typename BlockVector< value_type_ >::const_iterator
+BlockVector< value_type_ >::end() const
+{
+  return finish_;
+}
+
+template < typename value_type_ >
+inline void
+BlockVector< value_type_ >::push_back( const value_type_& value )
+{
+  // If this is the last element in the current block, add another block
+  if ( finish_.block_it_ == finish_.current_block_end_ - 1 )
+  {
+    blockmap_.emplace_back( max_block_size );
+  }
+  *finish_ = value;
+  ++finish_;
+}
+
+template < typename value_type_ >
+inline void
+BlockVector< value_type_ >::clear()
+{
+  for ( auto it = blockmap_.begin(); it != blockmap_.end(); ++it )
+  {
+    it->clear();
+  }
+  blockmap_.clear();
+  // Initialise the first block
+  blockmap_.emplace_back( max_block_size );
+  finish_ = begin();
+}
+
+template < typename value_type_ >
+inline size_t
+BlockVector< value_type_ >::size() const
+{
+  size_t element_index;
+  if ( finish_.block_index_ >= blockmap_.size() )
+  {
+    element_index = 0;
+  }
+  else
+  {
+    element_index =
+      finish_.block_it_ - blockmap_[ finish_.block_index_ ].begin();
+  }
+  return finish_.block_index_ * max_block_size + element_index;
+}
+
+template < typename value_type_ >
+inline typename BlockVector< value_type_ >::iterator
+BlockVector< value_type_ >::erase( const_iterator first, const_iterator last )
+{
+  assert( first.block_vector_ == this );
+  assert( last.block_vector_ == this );
+  assert( last < finish_ or last == finish_ );
+  // TODO: Except for some special cases, this method may kill performance as
```

Comment:
should these todos be addressed before merging?

Reply:
The function isn't that much used anyway, and making it more efficient would require quite a bit of mental work. So I removed the TODO.